staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator.go

```diff
@@ -18,298 +18,85 @@ package validatingadmissionpolicy
+	if f == v1.Ignore {
+		return cel.ActionAdmit
 	}
-	return &unstructured.Unstructured{Object: ret}, nil
+	return cel.ActionDeny
 }
 
-func objectToResolveVal(r runtime.Object) (interface{}, error) {
-	if r == nil || reflect.ValueOf(r).IsNil() {
-		return nil, nil
-	}
-	v, err := convertObjectToUnstructured(r)
-	if err != nil {
-		return nil, err
-	}
-	return v.Object, nil
-}
-
-func policyDecisionActionForError(f v1alpha1.FailurePolicyType) PolicyDecisionAction {
-	if f == v1alpha1.Ignore {
-		return ActionAdmit
-	}
-	return ActionDeny
-}
-
-// Validate validates all cel expressions in Validator and returns a PolicyDecision for each CEL expression or returns an error.
-// An error will be returned if failed to convert the object/oldObject/params/request to unstructured.
-// Each PolicyDecision will have a decision and a message.
-// policyDecision.message will be empty if the decision is allowed and no error met.
-func (v *CELValidator) Validate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) ([]PolicyDecision, error) {
-	// TODO: replace unstructured with ref.Val for CEL variables when native type support is available
-	decisions := make([]PolicyDecision, len(v.compilationResults))
-	var err error
-
-	oldObjectVal, err := objectToResolveVal(versionedAttr.VersionedOldObject)
-	if err != nil {
-		return nil, err
-	}
-	objectVal, err := objectToResolveVal(versionedAttr.VersionedObject)
-	if err != nil {
-		return nil, err
-	}
-	paramsVal, err := objectToResolveVal(versionedParams)
-	if err != nil {
-		return nil, err
+// Validate takes a list of Evaluation and a failure policy and converts them into actionable PolicyDecisions
+func (v *validator) Validate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) []cel.PolicyDecision {
+	var f v1.FailurePolicyType
+	if v.failPolicy == nil {
+		f = v1.Fail
+	} else {
+		f = *v.failPolicy
 	}
 
-	request := createAdmissionRequest(versionedAttr.Attributes)
-	requestVal, err := convertObjectToUnstructured(request)
+	evalResults, err := v.filter.ForInput(versionedAttr, versionedParams, cel.CreateAdmissionRequest(versionedAttr.Attributes))
 	if err != nil {
-		return nil, err
-	}
-	va := &validationActivation{
-		object:    objectVal,
-		oldObject: oldObjectVal,
-		params:    paramsVal,
-		request:   requestVal.Object,
-	}
-
-	var f v1alpha1.FailurePolicyType
-	if v.policy.Spec.FailurePolicy == nil {
-		f = v1alpha1.Fail
-	} else {
-		f = *v.policy.Spec.FailurePolicy
+		return []cel.PolicyDecision{
+			{
+				Action:     policyDecisionActionForError(f),
+				Evaluation: cel.EvalError,
+				Message:    err.Error(),
+			},
+		}
 	}
+	decisions := make([]cel.PolicyDecision, len(evalResults))
 
-	for i, compilationResult := range v.compilationResults {
-		validation := v.policy.Spec.Validations[i]
-
-		var policyDecision = &decisions[i]
+	for i, evalResult := range evalResults {
+		var decision = &decisions[i]
```

Comment:
My one non-nit, non-re-locate-this-code comment:

This implies that validations and evalResults have the same indexing order. This seems workable but a bit unwieldy.

Would it be possible for `EvaluationResult` to embed the original `ExpressionAccessor` object it is associated with? That way we don't need to worry about the internals of these two packages drifting.

We can also avoid caching `validations` in this object by doing so.